File: Task/Legendre-prime-counting-function/JavaScript/legendre-prime-counting-function-1.js
"use strict";

function countPrimesTo(lmt) {
  if (lmt < 3) { if (lmt < 2) return 0; else return 1; }
  const sqrtlmt = Math.sqrt(lmt) >>> 0;
  const oprms = function() {
    const mxndx = (sqrtlmt - 3) >>> 1;
    const arr = new Float64Array(mxndx + 1);
    for (let i = 0 >>> 0; i <= mxndx; ++i) arr[i] = (i + i + 3) >>> 0;
    let bp = 3 >>> 0;
    while (true) {
      let i = (bp - 3) >>> 1; let sqri = ((i + i) * (i + 3) + 3) >>> 0;
      if (sqri > mxndx) break;
      if (arr[i] != 0) for (; sqri <= mxndx; sqri += bp) arr[sqri] = 0;
      bp += 2;
    }
    return arr.filter(v => v != 0);
  }();
  function phi(x, a) {
    if (a <= 0) return x - Math.trunc(x / 2);
    const na = (a - 1) >>> 0; const p = oprms[na];
    if (x <= p) return 1;
    return phi(x, na) - phi(Math.trunc(x / p), na);
  }
  return phi(lmt, oprms.length) + oprms.length;
}

const start = Date.now();
for (let i = 0; i <= 9; ++i) console.log(`π(10**${i}) =`, countPrimesTo(10**i));
const elpsd = Date.now() - start;
console.log("This took", elpsd, "milliseconds.")

File: Task/Legendre-prime-counting-function/F-Sharp/legendre-prime-counting-function-1.fs
let countPrimes (lmt: uint64) =
  if lmt < 3UL then (if lmt < 2UL then 0L else 1L) else
  let sqrtlmt = lmt |> float |> sqrt |> uint64
  let mxndx = (sqrtlmt - 3UL) / 2UL |> int
  let oprms =
    let cb = Array.init (mxndx + 1) <| fun i -> uint32 (i + i + 3)
    let rec loopi i =
      let sqri = (i + i) * (i + 3) + 3
      if sqri > mxndx then () else
      if cb.[i] = 0u then loopi (i + 1) else
      let bp = i + i + 3
      let rec cull c = if c > mxndx then () else cb.[c] <- 0u; cull (c + bp)
      cull sqri; loopi (i + 1)
    loopi 0; cb |> Array.filter ((<>) 0u)
  let rec phi x a =
    if a <= 0 then x - (x >>> 1) |> int64 else
    let na = a - 1 in let p = uint64 oprms.[na]
    if x <= p then 1L else phi x na - phi (x / p) na
  phi lmt oprms.Length + int64 oprms.Length

let strt = System.DateTime.Now.Ticks

{ 0 .. 9 } |> Seq.iter (fun i ->
  printfn "π(10**%d) = %d" i (countPrimes (uint64(10. ** i))))

let elpsd = (System.DateTime.Now.Ticks - strt) / 10000L
printfn "This took %d milliseconds." elpsd
